feat(fieldlabel)!: core-tokens implementation (CSS-102)

bernhard-adobe opened the pull request:
## Description
BREAKING CHANGE: This migrates the Field label component to core tokens.

https://jira.corp.adobe.com/browse/CSS-102
Web Component Implementation: https://github.com/adobe/spectrum-web-components/pull/2529 refactor(field-label): use core tokens #2529


## How and where has this been tested?
 - **How this was tested:** comparing http://localhost:3002/docs/fieldlabel.html with the https://opensource.adobe.com/spectrum-css/fieldlabel.html component before and after.
 - **Browser(s) and OS(s) this was tested with:** Chrome Canary Version 105.0.5148.0 (Official Build) canary (arm64)

## Screenshots
<img width="655" alt="Screen Shot 2022-07-13 at 17 28 53" src="https://user-images.githubusercontent.com/52184321/178853951-ef2a25c5-32c4-4d40-b6b5-04bd714aca6a.png">


## To-do list
<!-- Put an "x" to indicate you've done each of the following -->
- [ ] If my change impacts other components, I have tested to make sure they don't break.
- [ ] If my change impacts documentation, I have updated the documentation accordingly.
- [x] I have read the [CONTRIBUTING document](/.github/CONTRIBUTING.md).
- [ ] I have tested these changes in Windows High Contrast mode.
<!-- If this pull request isn't ready, add any remaining tasks here -->
- [ ] This pull request is ready to merge.

bernhard-adobe commented on components/fieldlabel/index.css:
```diff
@@ -10,47 +10,62 @@ OF ANY KIND, either express or implied. See the License for the specific languag
+  font-weight: var(--spectrum-font-weight-regular);
+
+  /* line heights coming from core-tokens do not calculate correctly to the previous line-heights */
+  line-height: calc(var(--spectrum-fieldlabel-line-height) * var(--spectrum-fieldlabel-font-size));
```
@pfulton some of those line-heights don't match exactly the previous line-heights.
Is this calc the correct way to get a line-height?

pfulton replied:
Good question! I'll need to dig into this to get you a better answer.

bernhard-adobe replied:
Sounds good, lets double check with design just to make sure this is intentional

lazd replied:
Yeah I'd avoid this `calc()` the values above should have the `calc`, if anything.

lazd replied:
That looks good to me! Is `CJK` as a prefix already an established convention? It seems odd to see capitalized letters, and I wonder if a suffix is more in line with what we have already. Your call.

bernhard-adobe replied:
> That looks good to me! Is `CJK` as a prefix already an established convention? It seems odd to see capitalized letters, and I wonder if a suffix is more in line with what we have already. Your call.

I think so - it's in the core tokens file. 
I think Jess would be the one that decides on the final token names.

bernhard-adobe commented on components/fieldlabel/index.css:
```diff
@@ -66,27 +81,29 @@ governing permissions and limitations under the License.
-  --spectrum-tableform-labelsabove-item-gap: var(--spectrum-global-dimension-size-100);
+  /* for /docs/form.html to set field-label inside form */
+  --spectrum-tableform-border-spacing: 0 var(--spectrum-spacing-300);
+  /* matching 20px, missing global token for 20px */
```
@pfulton  We are missing 20px from the core-tokens. Is it ok to get 20px with a core-tokens addition?

pfulton replied:
Yeah, I think so. This will be something we'll need to keep an eye on, as I think I'd much rather have it coming as a definition from core tokens, but in the meantime (or if we have to), let's do this for now with an eye on possibly defining it ourselves in the `custom-*` files at some future point.

lazd replied:
If we're missing 20px in spacing values, then perhaps the component was designed wrong (or core tokens needs to have a new spacing value). We shouldn't be calcing this.

bernhard-adobe commented on components/fieldlabel/skin.css:
```diff
@@ -10,7 +10,7 @@ OF ANY KIND, either express or implied. See the License for the specific languag
 */
 
-.spectrum-FieldLabel,
+/* .spectrum-FieldLabel,
```
@pfulton  should I use here core tokens colors? Or is the inheritance here ok?

pfulton replied:
We should use the core tokens here. Also, we need to delete this `skin.css` file, and put the color stuff into the `index.css`

bernhard-adobe commented on components/tokens/custom-spectrum/custom-vars.css:
```diff
@@ -25,4 +26,34 @@ governing permissions and limitations under the License.
+  --spectrum-font-weight-extra-bold: 800;
+  --spectrum-font-weight-black: 900;
+
+  /* temporary hard-coding the font-sizes until we have core-tokens */
```
@pfulton  I am still not sure where to get those weights. I think it's still WIP for the core-tokens.

bernhard-adobe replied:
Still waiting for the core-tokens but we have those for now temporarily here.

pfulton reviewed (changes requested):
This is looking really great, @bernhard-adobe! Thanks so much for your work on this. I left a few comments here, and I think I answered the questions that you had left for me.

Two more things:
1. Let's be sure to add the `--mod-` prefixed custom properties in the places where we're using the other core tokens properties (look at ActionButton as your guide on this).
2. Let's be sure to add the `--highcontrast` prefixed custom properties in any places where we need high contrast mode to take effect.

bernhard-adobe commented on components/fieldlabel/index.css:
```diff
@@ -10,47 +10,62 @@ OF ANY KIND, either express or implied. See the License for the specific languag
+  --spectrum-fieldlabel-top-to-text: var(--spectrum-component-top-to-text-75);
+  --spectrum-fieldlabel-bottom-to-text: var(--spectrum-component-bottom-to-text-75);
+  --spectrum-fieldlabel-font-size: var(--spectrum-font-size-75);
+  --spectrum-fieldlabel-line-height: var(--spectrum-line-height-200);
```
I noticed in the inspection of the current docs site that the line-height changes.
size S and M = 15.6px
size L = 18.2px
Size XL = 20.8px

<img width="533" alt="Screen Shot 2022-07-18 at 15 33 00" src="https://user-images.githubusercontent.com/52184321/179621422-01b478e1-9595-4c04-925f-4dcef70b0fcf.png">
<img width="552" alt="Screen Shot 2022-07-18 at 15 33 11" src="https://user-images.githubusercontent.com/52184321/179621434-d449c94d-12ab-440d-806e-90a4592e9ab8.png">
<img width="237" alt="Screen Shot 2022-07-18 at 15 33 16" src="https://user-images.githubusercontent.com/52184321/179621454-f9f76da5-5cfd-4eb5-ba95-383e03404212.png">

I think from a visual perspective you do want to increase the line-heights for t-shirt sizing to keep the relational "growth" in sizing.

bernhard-adobe replied:
@pfulton I added the `--mod` modifiers. As there is no color-settings here I don't think we need to set `--highcontrast `

bernhard-adobe commented on components/fieldlabel/index.css:
```diff
@@ -10,47 +10,62 @@ OF ANY KIND, either express or implied. See the License for the specific languag
+  --spectrum-fieldlabel-top-to-text: var(--spectrum-component-top-to-text-100);
+  --spectrum-fieldlabel-bottom-to-text: var(--spectrum-component-bottom-to-text-100);
+  --spectrum-fieldlabel-font-size: var(--spectrum-font-size-100);
+  --spectrum-fieldlabel-line-height: var(--spectrum-CJK-line-height-100);
```
Will add! Good feedback, thank you

bernhard-adobe replied:
added to line 99

```CSS
/* international lang support */
.spectrum-FieldLabel {
  &:lang(ja),
  &:lang(zh),
  &:lang(ko) {
    line-height: var(--mod-fieldlabel-CJK-line-height, var(--spectrum-fieldlabel-CJK-line-height));
  }
}

bernhard-adobe commented on components/fieldlabel/index.css:
```diff
@@ -10,47 +10,62 @@ OF ANY KIND, either express or implied. See the License for the specific languag
+  --spectrum-fieldlabel-top-to-text: var(--spectrum-component-top-to-text-200);
+  --spectrum-fieldlabel-bottom-to-text: var(--spectrum-component-bottom-to-text-200);
+  --spectrum-fieldlabel-font-size: var(--spectrum-font-size-200);
+  --spectrum-fieldlabel-line-height: var(--spectrum-CJK-line-height-200);
```
same, lets double-check, maybe in a design workshop. I think we will discuss soon again the core-tokens for font-naming and weight and this might be a good additional question.

lazd commented on components/fieldlabel/index.css:
```diff
@@ -117,7 +134,7 @@ governing permissions and limitations under the License.
 
     & + .spectrum-Form-item {
-      margin-block-start: var(--spectrum-tableform-labelsabove-item-gap);
+      margin-block-start: var(--spectrum-spacing-100);
```
Any reason why this little fella isn't a component-level token above? It's not forbidden, per se, but I'd rather see it defined above for consistency.

bernhard-adobe replied:
@lazd I am actually not sure which item to use here. 
Spacing 100 would be 
```JSON
  "spacing-100": {
    "value": "8px"
  },
  ```
  
  Maybe use this? I don't see any other component level tokens for field-label matching 8px.

```JSON
  "field-label-top-to-asterisk-small": {
    "sets": {
      "desktop": {
        "value": "8px"
      },
      "mobile": {
        "value": "11px"
      }
    }
  },
  ```

lazd replied:
I think you just need to define that as a component-level token in CSS:

```
.spectrum--medium {
  --spectrum-field-label-top-to-asterisk-small: 8px;
}
.spectrum--large {
  --spectrum-field-label-top-to-asterisk-small: 11px;
}
```

bernhard-adobe replied:
Added, thanks @lazd

pfulton commented on components/fieldlabel/package.json:
```diff
@@ -17,22 +17,22 @@
   "peerDependencies": {
     "@spectrum-css/icon": "^3.0.21",
-    "@spectrum-css/vars": "^8.0.0"
+    "@spectrum-css/tokens": "^1.0.1"
```
Let's bump this to the most recent release: 

```suggestion
    "@spectrum-css/tokens": "^1.0.4"
```

bernhard-adobe replied:
bumped up

pfulton commented on components/fieldlabel/package.json:
```diff
@@ -17,22 +17,22 @@
     "@spectrum-css/stepper": "^3.0.26",
     "@spectrum-css/textfield": "^3.2.4",
-    "@spectrum-css/vars": "^8.0.0",
+    "@spectrum-css/tokens": "^1.0.1",
```
```suggestion
    "@spectrum-css/tokens": "^1.0.4",
```

pfulton commented on components/fieldlabel/index.css:
```diff
@@ -59,39 +95,49 @@ governing permissions and limitations under the License.
+  /* for /docs/form.html to set field-label inside form */
+  --spectrum-tableform-border-spacing: 0 var(--spectrum-spacing-300);
+  /* matching 20px, missing global token for 20px */
+  --spectrum-tableform-margin-calc: calc(var(--mod--spacing-300, var(--spectrum-spacing-300)) + var(--mod--spacing-300, var(--spectrum-spacing-200)));
```
This is a place where we won't want to define the `--mod-` custom properties. Instead, we'd define them later when we actually use `--spectrum-tableform-margin-calc`. Does that make sense?

bernhard-adobe replied:
simplified!

pfulton commented on components/fieldlabel/index.css:
```diff
@@ -10,47 +10,83 @@ OF ANY KIND, either express or implied. See the License for the specific languag
+  --spectrum-fieldlabel-font-size: var(--spectrum-font-size-75);
+
+  --spectrum-fieldlabel-line-height: var(--spectrum-line-height-100);
+  --spectrum-fieldlabel-CJK-line-height: var(--spectrum-CJK-line-height-100);
```
I'm trying to catch as many of these as I can, and I know that the core tokens value _is_ capitalized, but I think we should standardize on making it lowercase when we create our own aliases.

```suggestion
  --spectrum-fieldlabel-cjk-line-height: var(--spectrum-CJK-line-height-100);
```

bernhard-adobe replied:
made it as lower case as possible...

pfulton reviewed (changes requested):
A few things here @bernhard-adobe - thanks so much for your hard work on this and your patience in waiting for my review.

pfulton commented:
One more thing, @bernhard-adobe: the color for the `disabled` variant looks pretty different than it did before (see below with current and previous). Would you mind double-checking on that color?

<img width="1330" alt="disabled-fieldlabel" src="https://user-images.githubusercontent.com/360251/182642384-c9ce9935-5d27-45bd-93c9-2c0eb0de1608.png">

bernhard-adobe commented:
> One more thing, @bernhard-adobe: the color for the `disabled` variant looks pretty different than it did before (see below with current and previous). Would you mind double-checking on that color?
> 
> <img alt="disabled-fieldlabel" width="1330" src="https://user-images.githubusercontent.com/360251/182642384-c9ce9935-5d27-45bd-93c9-2c0eb0de1608.png">

That is indeed missing as it was not documented in the "Components Batch 1.xd" file. I guess the color would be `(--spectrum-disabled-content-color)` like all the other disabled components.

bernhard-adobe commented:
> > One more thing, @bernhard-adobe: the color for the `disabled` variant looks pretty different than it did before (see below with current and previous). Would you mind double-checking on that color?
> > <img alt="disabled-fieldlabel" width="1330" src="https://user-images.githubusercontent.com/360251/182642384-c9ce9935-5d27-45bd-93c9-2c0eb0de1608.png">
> 
> That is indeed missing as it was not documented in the "Components Batch 1.xd" file. I guess the color would be `(--spectrum-disabled-content-color)` like all the other disabled components.
<img width="309" alt="Screen Shot 2022-08-11 at 16 17 12" src="https://user-images.githubusercontent.com/52184321/184251917-891eceb4-6311-4261-9f40-344102c3d994.png">

bernhard-adobe commented:
rebuild Jenkins: https://spectrum-visual-regression.ci.corp.adobe.com/view/Spectrum%20CSS/job/css-vrt-test/37/

VRT results are: https://spectrum-visual-regression.ci.corp.adobe.com/view/Spectrum%20CSS/job/css-vrt-test/37/artifact/backstop_data/html_report/index.html

bernhard-adobe commented:
@pfulton color changes (as expected) for disabled and line-height changes (as expected):
<img width="311" alt="Screen Shot 2022-08-11 at 16 35 09" src="https://user-images.githubusercontent.com/52184321/184253917-0c497ee2-d2d9-43b6-9bc4-7b4fc4fafbef.png">

pfulton commented:
Released beta:
`5.0.0-beta.0`

pfulton commented:
New beta release:
`5.0.0-beta.1`

pfulton commented:
New release with syntax fix:
`5.0.0-beta.2`

bernhard-adobe commented on components/fieldlabel/index.css:
```diff
@@ -59,50 +98,74 @@ governing permissions and limitations under the License.
   display: inline-block;
-  padding-block: var(--spectrum-fieldlabel-m-side-padding-top) 0;
-  padding-inline: 0 var(--spectrum-fieldlabel-m-side-padding-right);
+  padding-block: var(--mod-fieldlabel-side-padding-top, var(--spectrum-fieldlabel-side-padding-top)) 0;
```
This seems to be a problem for Web components, the padding might be too small. 
![2022-09-22_14-50-55 (1)](https://user-images.githubusercontent.com/52184321/192054693-ba6ba429-3449-4a20-b253-5423ce561ff7.gif)

bernhard-adobe commented:
Fixed misaligned Field-Label padding for variant Side-Label.

@pfulton can you please publish another beta-release for the FieldLabel so I can pull in the latest changes in Web Component and re-run VRT over there? 

**VRT Results for Spectrum CSS are:** 
https://spectrum-visual-regression.ci.corp.adobe.com/view/Spectrum%20CSS/job/css-vrt-test/44/artifact/backstop_data/html_report/index.html

Line-height changed for FieldLabel. We have confirmed before that this is expected behavior.
With the latest push the Side-Label padding top  doesn't change anymore. This is in contrast with the documentation from the "Components batch 1.xd" file that says it should:
<img width="944" alt="Screen Shot 2022-09-25 at 18 47 33" src="https://user-images.githubusercontent.com/52184321/192173969-8dd84f5b-9ef2-42c6-9459-94edf21c9582.png">

If this is an intended change we should roll-back my last commit:  044d8ad1f96ece84dadbbd1fb1877bfec7b18aca

<img width="309" alt="Screen Shot 2022-09-25 at 18 27 13" src="https://user-images.githubusercontent.com/52184321/192173021-3cba977f-5073-4561-8e98-b37c40218fab.png">



Furthermore I aligned the CSS FieldLabel with the Web Components FieldLabel and the Line-Heights now match the Web Components. 

Because we changed the line-heights from previously 15.6 px to now 18px (no more sub-pixels) the position of the Label changes a bit. 
We may want to check in with design if this is intend or if we need to adjust the top-padding for the Side-Label variant.

 
<img width="1473" alt="Screen Shot 2022-09-25 at 18 43 42" src="https://user-images.githubusercontent.com/52184321/192174100-6a74d8a5-cefc-42f3-a701-0a1c2ced3db2.png">

<img width="1801" alt="Screen Shot 2022-09-25 at 18 42 06" src="https://user-images.githubusercontent.com/52184321/192174122-caf096f8-040a-4383-98d3-8c5a8e3801cd.png">
<img width="1603" alt="Screen Shot 2022-09-25 at 18 38 35" src="https://user-images.githubusercontent.com/52184321/192174125-9a75f57b-98fc-406c-aaf7-591e36b67a7d.png">

Please let me know @pfulton  if this sounds good. Thank you. 


(there are VRT changes with Switch and Menu reported as well, I think the baseline-images haven not been updates)

pfulton commented:
Released: `@spectrum-css/fieldlabel@5.0.0-beta.3`
